Lib/asyncio/tasks.py

```diff
@@ -433,15 +433,12 @@ async def wait_for(fut, timeout):
         try:
             await waiter
         except exceptions.CancelledError:
-            if fut.done():
-                return fut.result()
-            else:
-                fut.remove_done_callback(cb)
-                # We must ensure that the task is not running
-                # after wait_for() returns.
-                # See https://bugs.python.org/issue32751
-                await _cancel_and_wait(fut, loop=loop)
-                raise
+            fut.remove_done_callback(cb)
+            # We must ensure that the task is not running
+            # after wait_for() returns.
+            # See https://bugs.python.org/issue32751
+            await _cancel_and_wait(fut, loop=loop)
+            raise
```

Comment:
I think that the state `wait_for` is in right now (outside this PR) is as bad as
in swallowing a `KeyboardInterrupt` in a utility function. It makes it unusable
in my POV. There is also no way for user code to counteract, whereas the
resource leakage could actually be worked around.

This leads me to **possibility 1** how I think this issue could be solved:

Merge this PR, embrace the quirk and document it well.  
Tell the user something along the lines of
> Cancelling `wait_for` will leave the waited for future/coroutine in any
> state between (including both) not started and finished. Make sure to
> provide any necessary clean-up on cancellation.
> ```python
> try:
>      await asyncio.wait_for(foo, 123)
>  except asyncio.CancelledError:
>      ...  # do clean-up
>      raise
> ```
Or alternatively:    
> ```python
> try:
>      future = asyncio.ensure_future(foo) 
>      await asyncio.wait_for(future, 123)
>  except asyncio.CancelledError:
>      if fut.done() and not fut.cancelled() and fut.exception() is None:
>          ...  # the future completed, yet, we got cancelled
>      raise
> ```

**Possibility 2** I see is to not dismiss, but to defer the cancellation:
```suggestion
            if fut.done():
                # We got canclled, but we are already done. Therefore, we defer the
                # cancellation until the task yields to the event loop the next time.
                current_task().cancel() 
                return fut.result()
            else:
                fut.remove_done_callback(cb)
                # We must ensure that the task is not running
                # after wait_for() returns.
                # See https://bugs.python.org/issue32751
                await _cancel_and_wait(fut, loop=loop)
                raise
```

Reply:
To amend to that: this is the issue that I want to be solved (using pytest here, but it should get the point across):

```python
@pytest.mark.asyncio
async def test_wait_for_cancellation_propagates():
    inner = asyncio.get_event_loop().create_future()
    outer = asyncio.wait_for(inner, 0.1)
    outer_future = asyncio.ensure_future(outer)
    await asyncio.sleep(0)
    inner.set_result(None)  # inner is done
    outer_future.cancel()  # AND outer got cancelled

    # this fails starting with Python 3.8
    with pytest.raises(asyncio.CancelledError):
        await outer_future
```